### ISSUE TITLE
Have you tried different mask threshold?

### ISSUE THREAD

**Issue by laisimiao** [yikaiw/TokenFusion#3]
In 
https://github.com/yikaiw/TokenFusion/blob/6fe364ee6229e2d950dcadd9c011b5a48f45c0ec/semantic_segmentation/models/mix_transformer.py#L120
and in paper, the threshold is set to 0.02, but it seems to be a hyper-parameter of training? Have you tried tune it or where do you borrow from?

**Comment by yikaiw** [comment]
It is a hyper-parameter of training. This threshold simply follows the default threshold setting in [CEN](https://github.com/yikaiw/CEN) and I find the result already reaches SOTA.

**Comment by laisimiao** [comment]
I am just confused it. We learn a predictor like in DynamicViT by inputting tokens and outputting tokens' mask as indicator to reflect informative or uninformative ones. But it seems no direct supervision to guide learning of predictor and use final nll loss to propogate gradients to predictor module so that no extract meaning about predictor function?

**Comment by yikaiw** [comment]
> ok. but when you write code of CEN or something ealier, you should investigate the value of activation map and set to 0.02 right?

In CEN (appendix), we investigate the value of threshold and observe that the performance is not sensitive given the threshold around 0.02.

**Comment by yikaiw** [comment]
> I am just confused it. We learn a predictor like in DynamicViT by inputting tokens and outputting tokens' mask as indicator to reflect informative or uninformative ones. But it seems no direct supervision to guide learning of predictor and use final nll loss to propogate gradients to predictor module so that no extract meaning about predictor function?

Yes, currently there is no direct supervision added to the predictor. It is hoped that when l_1 norm is placed along the gradient path (of the final nll loss), it can find out uninformative tokens during training.

**Comment by laisimiao** [comment]
Thanks for your reply. Please help me figure out anothor two questions:
1. in paper equation5, there is l1 loss for scoring function (also called predictor in code) but no implementation in code, right?
2. in line 139, you use logsoftmax to get mask , but the output of logsoftmax is negative number, I think they won't be larger than the mask threshold 0.02 forever?https://github.com/yikaiw/TokenFusion/blob/9539baff20420926d8fff7d2b54646b9cd8d7602/semantic_segmentation/models/mix_transformer.py#L125-L140

**Comment by laisimiao** [comment]
Oh for first question, I have found that part code.

**Comment by yikaiw** [comment]
The mask is not directly obtained by logsoftmax, but uses a softmax function to force the output positive:

https://github.com/yikaiw/TokenFusion/blob/9539baff20420926d8fff7d2b54646b9cd8d7602/semantic_segmentation/models/mix_transformer.py#L361-L366
